File: Day #1059/solution.py
# Day 1059: Word sense disambiguation: for each ambiguous word pick the meaning whose words
# overlap most with the sentence context (other words). Tie-break -> first meaning.
# Time: O(W * M * L), Space: O(L) for the context set.
import re


def tokenize(s):
    return [t for t in re.split(r"[^a-z0-9]+", s.lower()) if t]


def disambiguate(dictionary, sentence):
    words = tokenize(sentence)
    results = {}
    for i, w in enumerate(words):
        meanings = dictionary.get(w)
        if not meanings or len(meanings) <= 1:
            continue
        context = set(words[:i] + words[i + 1:])
        best, best_score = 0, -1
        for m, meaning in enumerate(meanings):
            score = sum(1 for t in tokenize(meaning) if t in context)
            if score > best_score:
                best_score, best = score, m
        results[w] = meanings[best]
    return results


if __name__ == "__main__":
    dictionary = {
        "bank": [
            "financial institution where people deposit money",
            "land beside a river or lake",
        ]
    }
    sentence = "I went to get money from the bank"
    for word, meaning in disambiguate(dictionary, sentence).items():
        print(f"{word}: {meaning}")

File: Day #1059/solution.js
// Word sense disambiguation: for each ambiguous word pick the meaning whose words
// overlap most with the sentence context (other words). Tie-break -> first meaning.
// Time: O(W * M * L), Space: O(L) for the context set.
"use strict";

function tokenize(s) {
  return s.toLowerCase().split(/[^a-z0-9]+/).filter(Boolean);
}

function disambiguate(dict, sentence) {
  const words = tokenize(sentence);
  const results = {};
  for (let i = 0; i < words.length; i++) {
    const w = words[i];
    const meanings = dict[w];
    if (!meanings || meanings.length <= 1) continue;

    const context = new Set(words.filter((_, j) => j !== i));
    let best = 0, bestScore = -1;
    for (let m = 0; m < meanings.length; m++) {
      let score = 0;
      for (const t of tokenize(meanings[m])) if (context.has(t)) score++;
      if (score > bestScore) { bestScore = score; best = m; }
    }
    results[w] = meanings[best];
  }
  return results;
}

const dict = {
  bank: [
    "financial institution where people deposit money",
    "land beside a river or lake",
  ],
};
const sentence = "I went to get money from the bank";
for (const [word, meaning] of Object.entries(disambiguate(dict, sentence))) {
  console.log(`${word}: ${meaning}`);
}
